File: Britefury/CommandHistory/CommandHistory.py
##-*************************
##-* This program is free software; you can use it, redistribute it and/or modify it
##-* under the terms of the GNU General Public License version 2 as published by the
##-* Free Software Foundation. The full text of the GNU General Public License
##-* version 2 can be found in the file named 'COPYING' that accompanies this
##-* program. This source code is (C)copyright Geoffrey French 1999-2007.
##-*************************
from Britefury.Util.SignalSlot import ClassSignal


class Command (object):
	def __init__(self):
		self._bFinished = False


	def execute(self):
		"Execute the command"
		assert False, 'abstract'

	def unexecute(self):
		"Unexecute the command"
		assert False, 'abstract'


	def canJoinWith(self, command):
		"Determine if @command can be joined with @self"
		return False

	def joinWith(self, command):
		"Join @command with @self"
		assert False, 'abstract'




class CommandHistory (object):
	changedSignal = ClassSignal()


	def __init__(self):
		self._past = []
		self._future = []
		self._bCommandsBlocked = False
		self._bFrozen = False
		self._freezeCount = 0
		self._trackers = {}
		self._trackedObjects = set()


	def addCommand(self, command):
		"Add a command to the history"
		assert isinstance( command, Command ), 'command is not an instance of UndoStackAction'
		if not self._bCommandsBlocked:
			self._future = []
			top = self.top()
			# Attempt to join @command with @top
			if top is not None  and  top.canJoinWith( command ):
				top.joinWith( command )
			else:
				if self._bFrozen:
					self._past[-1].append( command )
				else:
					self._past.append( command )
			self.changedSignal.emit( self )


	def undo(self):
		"Undo a command"
		if self._bFrozen:
			self.thaw()
		command = self._past.pop()
		if isinstance( command, list ):
			for c in reversed( command ):
				self._p_unexecuteCommand( c )
		else:
			self._p_unexecuteCommand( command )
		self._future.append( command )
		self.changedSignal.emit( self )

	def redo(self):
		"Redo a command"
		if self._bFrozen:
			self.thaw()
		command = self._future.pop()
		if isinstance( command, list ):
			for c in command:
				self._p_executeCommand( c )
		else:
			self._p_executeCommand( command )
		self._past.append( command )
		self.changedSignal.emit( self )


	def clear(self):
		"Clear the command history"
		self._past = []
		self._future = []
		self.changedSignal.emit( self )



	def freeze(self):
		"Freeze the command history"
		if not self._bFrozen:
			if self.top() != []:
				self._past.append( [] )
		self._bFrozen = True
		self._freezeCount += 1

	def freezeJoin(self):
		"Freeze the command history, joining onto previous command"
		if len( self._past ) > 0:
			if not isinstance( self._past[-1], list ):
				self._past[-1] = [ self._past[-1] ]
			self._bFrozen = True
			self._freezeCount += 1
		else:
			self.freeze()

	def thaw(self):
		"Thaw the command history"
		self._freezeCount -= 1
		if self._freezeCount == 0:
			if len( self._past[-1] ) == 0:
				self._past.pop()
			self._bFrozen = False

		self._freezeCount = max( self._freezeCount, 0 )



	def top(self):
		"Get the top command"
		if len( self._past ) > 0:
			top = self._past[-1]
			if isinstance( top, list ):
				if len( top ) > 0:
					return top[-1]
				else:
					return None
			else:
				return top
		else:
			return None


	def finishCommand(self):
		"Disable command joining"
		top = self.top()
		if top is not None:
			top._bFinished = True




	def canUndo(self):
		return len( self._past )  >  0

	def canRedo(self):
		return len( self._future )  >  0

	def getNumUndoCommands(self):
		"Get the number of actions that can be undone"
		return len( self._past )

	def getNumRedoCommands(self):
		"Get the number of actions that can be redone"
		return len( self._future )




	def track(self, obj):
		try:
			trackerClass = obj.trackerClass
		except AttributeError:
			raise TypeError, 'object \'%s\' is not trackable; \'trackerClass\' attribute not found' % ( obj, )

		if trackerClass is None:
			raise TypeError, 'trackable object \'%s\' has no tracker class' % ( obj, )
		if not isinstance( trackerClass, type ):
			raise TypeError, 'tracker class must be a type, it is \'%s\'' % ( trackerClass, )

		if obj not in self._trackedObjects:
			try:
				tracker = self._trackers[obj.trackerClass]
			except KeyError:
				tracker = obj.trackerClass( self )
				self._trackers[obj.trackerClass] = tracker

			self._trackedObjects.add( obj )
			tracker.track( obj )


	def stopTracking(self, obj):
		try:
			trackerClass = obj.trackerClass
		except AttributeError:
			raise TypeError, 'object \'%s\' is not trackable; \'trackerClass\' attribute not found' % ( obj, )

		if trackerClass is None:
			raise TypeError, 'trackable object \'%s\' has no tracker class' % ( obj, )
		if not isinstance( trackerClass, type ):
			raise TypeError, 'tracker class must be a type, it is \'%s\'' % ( trackerClass, )

		if obj in self._trackedObjects:
			try:
				tracker = self._trackers[obj.trackerClass]
			except KeyError:
				assert False, 'no tracker'

			self._trackedObjects.remove( obj )
			tracker.stopTracking( obj )



	def _p_blockCommands(self):
		self._bCommandsBlocked = True

	def _p_unblockCommands(self):
		self._bCommandsBlocked = False



	def _p_executeCommand(self, command):
		self._p_blockCommands()
		command.execute()
		self._p_unblockCommands()

	def _p_unexecuteCommand(self, command):
		self._p_blockCommands()
		command.unexecute()
		self._p_unblockCommands()







# *************************** UNIT TESTING ***************************

if __name__ == '__main__':
	import unittest

	class TestCommand (Command):
		def __init__(self, target, value):
			super( TestCommand, self ).__init__()
			self._target = target
			self._value = value
			self._oldValue = None

		def execute(self):
			self._oldValue = self._target[0]
			self._target[0] = self._value

		def unexecute(self):
			self._target[0] = self._oldValue


	class CommandHistoryTest (unittest.TestCase):
		def testStack(self):
			target = [ 0 ]
			stack = CommandHistory()

			self.assertEqual( target[0], 0 )
			self.assertEqual( stack.getNumUndoCommands(), 0 )
			self.assertEqual( stack.getNumRedoCommands(), 0 )

			stack.addCommandAndExecute( TestCommand( target, 1 ) )
			self.assertEqual( target[0], 1 )
			self.assertEqual( stack.getNumUndoCommands(), 1 )
			self.assertEqual( stack.getNumRedoCommands(), 0 )

			stack.addCommandAndExecute( TestCommand( target, 2 ) )
			self.assertEqual( target[0], 2 )
			self.assertEqual( stack.getNumUndoCommands(), 2 )
			self.assertEqual( stack.getNumRedoCommands(), 0 )

			stack.addCommandAndExecute( TestCommand( target, 3 ) )
			self.assertEqual( target[0], 3 )
			self.assertEqual( stack.getNumUndoCommands(), 3 )
			self.assertEqual( stack.getNumRedoCommands(), 0 )

			stack.addCommandAndExecute( TestCommand( target, 4 ) )
			self.assertEqual( target[0], 4 )
			self.assertEqual( stack.getNumUndoCommands(), 4 )
			self.assertEqual( stack.getNumRedoCommands(), 0 )

			stack.undo()
			self.assertEqual( target[0], 3 )
			self.assertEqual( stack.getNumUndoCommands(), 3 )
			self.assertEqual( stack.getNumRedoCommands(), 1 )

			stack.undo()
			self.assertEqual( target[0], 2 )
			self.assertEqual( stack.getNumUndoCommands(), 2 )
			self.assertEqual( stack.getNumRedoCommands(), 2 )

			stack.redo()
			self.assertEqual( target[0], 3 )
			self.assertEqual( stack.getNumUndoCommands(), 3 )
			self.assertEqual( stack.getNumRedoCommands(), 1 )

			stack.addCommandAndExecute( TestCommand( target, -1 ) )
			self.assertEqual( target[0], -1 )
			self.assertEqual( stack.getNumUndoCommands(), 4 )
			self.assertEqual( stack.getNumRedoCommands(), 0 )

			stack.undo()
			self.assertEqual( target[0], 3 )
			self.assertEqual( stack.getNumUndoCommands(), 3 )
			self.assertEqual( stack.getNumRedoCommands(), 1 )

			stack.undo()
			self.assertEqual( target[0], 2 )
			self.assertEqual( stack.getNumUndoCommands(), 2 )
			self.assertEqual( stack.getNumRedoCommands(), 2 )



	unittest.main()

File: java/BritefuryJ/CommandHistory/CommandHistory.java
//##* This program is free software; you can use it, redistribute it and/or modify it
//##* under the terms of the GNU General Public License version 2 as published by the
//##* Free Software Foundation. The full text of the GNU General Public License
//##* version 2 can be found in the file named 'COPYING' that accompanies this
//##* program. This source code is (C)copyright Geoffrey French 2008.
//##************************
package BritefuryJ.CommandHistory;

import java.awt.Color;
import java.util.ArrayList;

import BritefuryJ.AttributeTable.SimpleAttributeTable;
import BritefuryJ.DefaultPerspective.Presentable;
import BritefuryJ.DefaultPerspective.Pres.ObjectBox;
import BritefuryJ.DocPresent.Border.SolidBorder;
import BritefuryJ.DocPresent.Painter.FillPainter;
import BritefuryJ.IncrementalView.FragmentView;
import BritefuryJ.ObjectPresentation.PresentationStateListenerList;
import BritefuryJ.Pres.InnerFragment;
import BritefuryJ.Pres.Pres;
import BritefuryJ.Pres.Primitive.Arrow;
import BritefuryJ.Pres.Primitive.Border;
import BritefuryJ.Pres.Primitive.Column;
import BritefuryJ.Pres.Primitive.Label;
import BritefuryJ.Pres.Primitive.Primitive;
import BritefuryJ.Pres.Primitive.Row;
import BritefuryJ.StyleSheet.StyleSheet;

public class CommandHistory implements CommandHistoryController, Presentable
{
	private static abstract class Entry implements Presentable
	{
		public abstract void execute();
		public abstract void unexecute();
		public abstract boolean isEmpty();
		public abstract Command top();
	}
	
	private static class SingleEntry extends Entry
	{
		private Command command;
		
		public SingleEntry(Command command)
		{
			this.command = command;
		}
		
		public void execute()
		{
			command.execute();
		}

		public void unexecute()
		{
			command.unexecute();
		}
		
		public boolean isEmpty()
		{
			return false;
		}
		
		public Command top()
		{
			return command;
		}

		
		@Override
		public Pres present(FragmentView fragment, SimpleAttributeTable inheritedState)
		{
			return new InnerFragment( command );
		}
	}

	private static class MultiEntry extends Entry
	{
		private ArrayList<Command> commands;
		
		public MultiEntry()
		{
			commands = new ArrayList<Command>();
		}
		
		public void execute()
		{
			for (Command c: commands)
			{
				c.execute();
			}
		}

		public void unexecute()
		{
			for (int i = commands.size() - 1; i >= 0; i--)
			{
				commands.get( i ).unexecute();
			}
		}

		public boolean isEmpty()
		{
			return commands.size() == 0;
		}

		public Command top()
		{
			if ( commands.size() > 0 )
			{
				return commands.get( commands.size() - 1 );
			}
			else
			{
				return null;
			}
		}
		
		
		public void add(Command command)
		{
			commands.add( command );
		}

		
		@Override
		public Pres present(FragmentView fragment, SimpleAttributeTable inheritedState)
		{
			return new ObjectBox( "Command group", new Column( Pres.mapCoerce( commands ) ) );
		}
	}
	
	
	
	private ArrayList<Entry> past, future;
	private boolean bCommandsBlocked, bFrozen;
	private int freezeCount;
	private CommandHistoryListener listener;
	private PresentationStateListenerList presStateListeners = null;
	
	
	
	
	//
	//
	// Constructor
	//
	//
	public CommandHistory()
	{
		past = new ArrayList<Entry>();
		future = new ArrayList<Entry>();
		bCommandsBlocked = false;
		bFrozen = false;
		freezeCount = 0;
		listener = null;
	}
	
	
	
	public void setCommandHistoryListener(CommandHistoryListener listener)
	{
		this.listener = listener;
	}
	
	
	public void addCommand(Command command)
	{
		if ( !bCommandsBlocked )
		{
			future.clear();
			Command top = topCommand();
			// Attempt to merge @command into @top
			if ( top != null  &&  top.canMergeFrom( command ) )
			{
				top.mergeFrom( command );
			}
			else
			{
				if ( bFrozen )
				{
					((MultiEntry)past.get( past.size() - 1 )).add( command );
				}
				else
				{
					past.add( new SingleEntry( command ) );
				}
			}
			
			if ( listener != null )
			{
				listener.onCommandHistoryChanged( this );
			}
			
			onModified();
		}
	}
	
	
	
	public void undo()
	{
		if ( bFrozen )
		{
			thaw();
		}
		
		if ( past.size() >= 1 )
		{
			Entry entry = past.get( past.size() - 1 );
			past.remove( past.size() - 1 );
			unexecuteEntry( entry );
			future.add( entry );
			
			if ( listener != null )
			{
				listener.onCommandHistoryChanged( this );
			}

			onModified();
		}
	}
	
	
	public void redo()
	{
		if ( bFrozen )
		{
			thaw();
		}
		
		if ( future.size() >= 1 )
		{
			Entry entry = future.get( future.size() - 1 );
			future.remove( future.size() - 1 );
			executeEntry( entry );
			past.add( entry );
			
			if ( listener != null )
			{
				listener.onCommandHistoryChanged( this );
			}

			onModified();
		}
	}
	
	
	public void clear()
	{
		past.clear();
		future.clear();
		
		if ( listener != null )
		{
			listener.onCommandHistoryChanged( this );
		}
		
		onModified();
	}
	
	
	
	public void freeze()
	{
		if ( !bFrozen )
		{
			// Add a new empty multi-entry
			past.add( new MultiEntry() );
		}
		
		bFrozen = true;
		freezeCount++;
	}
	
	public void thaw()
	{
		freezeCount--;
		if ( freezeCount == 0 )
		{
			// If there is an existing top entry, and it is empty, remove it.
			Entry top = topEntry();
			if ( top != null  &&  top.isEmpty() )
			{
				past.remove( past.size() - 1 );
			}
			bFrozen = false;
		}
		
		freezeCount = Math.max( freezeCount, 0 );
	}
	
		
	public boolean canUndo()
	{
		return past.size() > 0;
	}

	public boolean canRedo()
	{
		return future.size() > 0;
	}

	
	public int getNumUndoCommands()
	{
		return past.size();
	}

	public int getNumRedoCommands()
	{
		return future.size();
	}
	
	
	
	public void track(Trackable t)
	{
		CommandHistory h = t.getCommandHistory();
		if ( h == null )
		{
			t.setCommandHistory( this );
			t.trackContents( this );
		}
		else if ( h != this )
		{
			throw new RuntimeException( "Trackable object is already being tracked by a different command history" );
		}
	}

	public void stopTracking(Trackable t)
	{
		t.stopTrackingContents( this );
		t.setCommandHistory( null );
	}
	
	
	public void track(Object x)
	{
		if ( x != null  &&  x instanceof Trackable )
		{
			track( (Trackable)x );
		}
	}

	public void stopTracking(Object x)
	{
		if ( x != null  &&  x instanceof Trackable )
		{
			stopTracking( (Trackable)x );
		}
	}
	
	
	
	
	protected Command topCommand()
	{
		if ( past.size() > 0 )
		{
			return past.get( past.size() - 1 ).top();
		}
		else
		{
			return null;
		}
	}
	
	protected Entry topEntry()
	{
		if ( past.size() > 0 )
		{
			return past.get( past.size() - 1 );
		}
		else
		{
			return null;
		}
	}
	
	
	
	
	
	
	private void blockCommands()
	{
		bCommandsBlocked = true;
	}
	
	private void unblockCommands()
	{
		bCommandsBlocked = false;
	}
	
	
	private void executeEntry(Entry entry)
	{
		blockCommands();
		entry.execute();
		unblockCommands();
	}

	private void unexecuteEntry(Entry entry)
	{
		blockCommands();
		entry.unexecute();
		unblockCommands();
	}
	
	
	private void onModified()
	{
		presStateListeners = PresentationStateListenerList.onPresentationStateChanged( presStateListeners, this );
	}



	@Override
	public Pres present(FragmentView fragment, SimpleAttributeTable inheritedState)
	{
		presStateListeners = PresentationStateListenerList.addListener( presStateListeners, fragment );

		Pres pastTitleTop = pastTitleStyle.applyTo( new Row( new Pres[] { new Arrow( Arrow.Direction.DOWN, 14.0 ).alignVCentre(), new Label( "Past" ) } ) );
		Pres pastContents = new Column( Pres.mapCoerce( past ) );
		Pres pastTitleBottom = pastTitleStyle.applyTo( new Row( new Pres[] { new Arrow( Arrow.Direction.UP, 14.0 ).alignVCentre(), new Label( "Past" ) } ) );
		Pres pastBox = pastBorderStyle.applyTo( new Border( listBoxStyle.applyTo( new Column( new Pres[] { pastTitleTop, pastContents, pastTitleBottom } ) ) ) );

		Pres futureTitleTop = futureTitleStyle.applyTo( new Row( new Pres[] { new Arrow( Arrow.Direction.DOWN, 14.0 ).alignVCentre(), new Label( "Future" ) } ) );
		Pres futureContents = new Column( Pres.mapCoerce( future ) );
		Pres futureTitleBottom = futureTitleStyle.applyTo( new Row( new Pres[] { new Arrow( Arrow.Direction.UP, 14.0 ).alignVCentre(), new Label( "Future" ) } ) );
		Pres futureBox = futureBorderStyle.applyTo( new Border( listBoxStyle.applyTo( new Column( new Pres[] { futureTitleTop, futureContents, futureTitleBottom } ) ) ) );
		
		Pres mainBox = commandHistoryColumnStyle.applyTo( new Column( new Pres[] { pastBox, futureBox } ) );
		
		return new ObjectBox( "CommandHistory", mainBox );
	}
	
	
	private static final StyleSheet pastTitleStyle = StyleSheet.instance.withAttr( Primitive.foreground, new Color( 0.5f, 0.0f, 0.5f ) )
			.withAttr( Primitive.shapePainter, new FillPainter( new Color( 0.5f, 0.0f, 0.5f ) ) )
			.withAttr( Primitive.fontFace, "Serif" ).withAttr( Primitive.fontSmallCaps, true );
	private static final StyleSheet futureTitleStyle = StyleSheet.instance.withAttr( Primitive.foreground, new Color( 0.0f, 0.25f, 0.5f ) )
			.withAttr( Primitive.shapePainter, new FillPainter( new Color( 0.0f, 0.25f, 0.5f ) ) )
			.withAttr( Primitive.fontFace, "Serif" ).withAttr( Primitive.fontSmallCaps, true );
	private static final StyleSheet pastBorderStyle = StyleSheet.instance.withAttr( Primitive.border, new SolidBorder( 2.0, 3.0, 10.0, 10.0, new Color( 0.5f, 0.0f, 0.5f ), null ) );
	private static final StyleSheet futureBorderStyle = StyleSheet.instance.withAttr( Primitive.border, new SolidBorder( 2.0, 3.0, 10.0, 10.0, new Color( 0.0f, 0.25f, 0.5f ), null ) );
	private static final StyleSheet listBoxStyle = StyleSheet.instance.withAttr( Primitive.columnSpacing, 10.0 );
	private static final StyleSheet commandHistoryColumnStyle = StyleSheet.instance.withAttr( Primitive.columnSpacing, 20.0 );
}
